### PR TITLE
Support basic auth in Prometheus

### LINKED ISSUE #12294: support basic auth in prometheus connector
Hi,

I'm trying to connect trino to our prometheus, but it's password protected [basic auth](https://prometheus.io/docs/guides/basic-auth/)

I find that the connector supports bearer token , but I can't find any information of that in prometheus document.

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

* Fixes #12294

## Documentation

(x) Sufficient documentation is included in this PR.

## Release notes

(x) Release notes entries required with the following suggested text:

```markdown
# Prometheus
* Add support for basic authentication. ({issue}`12302`)
```

### CODE REVIEW

File: plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/PrometheusClient.java [hunk 1]
```diff
@@ -66,9 +67,10 @@ public PrometheusClient(PrometheusConnectorConfig config, JsonCodec<Map<String,
         requireNonNull(metricCodec, "metricCodec is null");
         requireNonNull(typeManager, "typeManager is null");
 
-        httpClient = new Builder().readTimeout(Duration.ofMillis(config.getReadTimeout().toMillis())).build();
+        Builder httpClient = new Builder().readTimeout(Duration.ofMillis(config.getReadTimeout().toMillis()));
```
Comment (Praveen2112) [inline plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/PrometheusClient.java:72]: Can we rename it as `clientBuilder` ?

File: testing/trino-product-tests-launcher/src/main/resources/docker/presto-product-tests/conf/environment/multinode-all/prometheus.properties [hunk 2]
```diff
@@ -3,5 +3,5 @@ prometheus.uri=http://host1.invalid:9090
 prometheus.query.chunk.size.duration=1d
 prometheus.max.query.range.duration=21d
 prometheus.cache.ttl=30s
-prometheus.bearer.token.file=/path/to/bearer/token/file
```
Comment (Praveen2112) [inline testing/trino-product-tests-launcher/src/main/resources/docker/presto-product-tests/conf/environment/multinode-all/prometheus.properties:6]: Shouldn't we delete this file ?

File: plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/PrometheusClient.java [hunk 3]
```diff
@@ -164,4 +164,19 @@ private Optional<String> getBearerAuthInfoFromFile()
             }
         });
     }
+
+    private static void setupTokenAuth(OkHttpClient.Builder clientBuilder, Optional<String> accessToken)
+    {
+        if (accessToken.isPresent()) {
```
Comment (Praveen2112) [inline plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/PrometheusClient.java:180]: Can we move it to some functional representation 

```
        accessToken
                .map(PrometheusClient::tokenAuth)
                .ifPresent(clientBuilder::addInterceptor);
```

File: plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/PrometheusConnectorConfig.java [hunk 4]
```diff
@@ -107,6 +110,33 @@ public PrometheusConnectorConfig setBearerTokenFile(File bearerTokenFile)
...
+
+    @NotNull
+    public Optional<String> getPassword()
+    {
+        return Optional.ofNullable(password);
+    }
+
+    @Config("prometheus.auth.password")
+    @ConfigSecuritySensitive
+    public PrometheusConnectorConfig setPassword(String password)
```
Comment (Praveen2112) [inline plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/PrometheusConnectorConfig.java:134]: Can we add some validation like if the userName is configured then we require the password and vice versa ?

Comment (Praveen2112) [reply plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/PrometheusConnectorConfig.java:134]: Is it possible to have both the basic and token based authentication ?

File: plugin/trino-prometheus/src/test/java/io/trino/plugin/prometheus/TestPrometheusBasicAuth.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    @Override
+    protected QueryRunner createQueryRunner()
+            throws Exception
+    {
+        PrometheusServer server = closeAfterClass(new PrometheusServer(LATEST_VERSION));
+        return createPrometheusQueryRunner(server, ImmutableMap.of(), ImmutableMap.of("prometheus.auth.user", USER, "prometheus.auth.password", PASSWORD));
+    }
+
+    @Test
+    public void testSelect()
```
Comment (Praveen2112) [inline plugin/trino-prometheus/src/test/java/io/trino/plugin/prometheus/TestPrometheusBasicAuth.java:42]: Any additional tests for invalid credentials ?

File: plugin/trino-prometheus/src/test/java/io/trino/plugin/prometheus/PrometheusServer.java [hunk 6]
```diff
@@ -20,24 +20,40 @@
...
+        this(DEFAULT_VERSION);
+    }
+
+    public PrometheusServer(String version)
+    {
+        this.dockerContainer = new GenericContainer<>("prom/prometheus:" + version)
                 .withExposedPorts(PROMETHEUS_PORT)
                 .waitingFor(Wait.forHttp("/"))
                 .withStartupTimeout(Duration.ofSeconds(120));
+        if (version.equals(LATEST_VERSION)) {
```
Comment (Praveen2112) [inline plugin/trino-prometheus/src/test/java/io/trino/plugin/prometheus/PrometheusServer.java:51]: The basic auth is introduced only in the latest version i.e 2.35 ?

File: testing/trino-product-tests-launcher/src/main/resources/docker/presto-product-tests/conf/environment/multinode-all/prometheus.properties (hunk 2 above)
Comment (ebyhr) [reply testing/trino-product-tests-launcher/src/main/resources/docker/presto-product-tests/conf/environment/multinode-all/prometheus.properties:6]: It's modification, not delete.

File: plugin/trino-prometheus/src/test/java/io/trino/plugin/prometheus/PrometheusServer.java (hunk 6 above)
Comment (ebyhr) [reply plugin/trino-prometheus/src/test/java/io/trino/plugin/prometheus/PrometheusServer.java:51]: It was introduced in `v2.24.0`.

File: plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/PrometheusConnectorConfig.java (hunk 4 above)
Comment (ebyhr) [reply plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/PrometheusConnectorConfig.java:134]: Not sure if Prometheus supports two authentications. Added guard logic to allow only single auth.

**Comment by ebyhr** [comment]
@Praveen2112 Updated. Please take another look when you have time.

File: plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/PrometheusConnectorConfig.java [hunk 7]
```diff
@@ -129,5 +159,11 @@ public void checkConfig()
         if (maxQueryRangeDuration < queryChunkSizeDuration) {
             throw new ConfigurationException(ImmutableList.of(new Message("prometheus.max.query.range.duration must be greater than prometheus.query.chunk.size.duration")));
         }
+        if (getBearerTokenFile().isPresent() && (getUser().isPresent() || getPassword().isPresent())) {
+            throw new IllegalStateException("Either on of bearer token file or basic authentication should be used");
+        }
+        if ((getUser().isPresent() && getPassword().isEmpty()) || (getPassword().isPresent() && getUser().isEmpty())) {
```
Comment (Praveen2112) [inline plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/PrometheusConnectorConfig.java:165]: We could try to XOR between user and password right ?

**Review by Praveen2112 — APPROVED** [review]

File: plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/PrometheusConnectorConfig.java [hunk 8]
```diff
@@ -129,5 +159,11 @@ public void checkConfig()
         if (maxQueryRangeDuration < queryChunkSizeDuration) {
             throw new ConfigurationException(ImmutableList.of(new Message("prometheus.max.query.range.duration must be greater than prometheus.query.chunk.size.duration")));
         }
+        if (getBearerTokenFile().isPresent() && (getUser().isPresent() || getPassword().isPresent())) {
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/PrometheusConnectorConfig.java:162]: Let me skip addressing this comment to minimize the change.

**Pull request merged**